Step2

yesjehmi opened the pull request.

elie-jehmi committed:
login & refactoring 적용.

elie-jehmi committed:
Question step2 구현.
1. QuestionRepository 인터페이스 추가.
2. QnaService TODO 구현.
3. HomeController QnaList 조회 기능 추가.
4. LoginAcceptanceTest 추가
4-1. LoginAcceptance 기반으로 LoginController추가
5. QnaAcceptanceTest 추가
5-1. QnaAcceptance 기반으로 QnaController추가
6. UserController 불필요 path 제거

yesjehmi committed:
Merge branch 'yesjehmi' into step2

javajigi commented on src/main/java/nextstep/domain/QuestionRepository.java:
```diff
@@ -2,6 +2,11 @@
+
 public interface QuestionRepository extends JpaRepository<Question, Long> {
     Iterable<Question> findByDeleted(boolean deleted);
+    Optional<Question> findFirstByTitleAndContents(String title, String contents);
```
테스트를 위한 이 메소드 없이 구현 진행해 본다.

javajigi commented on src/main/java/nextstep/domain/QuestionRepository.java:
```diff
@@ -2,6 +2,11 @@
 public interface QuestionRepository extends JpaRepository<Question, Long> {
     Iterable<Question> findByDeleted(boolean deleted);
+    Optional<Question> findFirstByTitleAndContents(String title, String contents);
+    Optional<Question> findByIdAndWriter(Long id, User writer);
```
이 메소드 없이 구현 진행해 본다.

javajigi commented on src/main/java/nextstep/domain/QuestionRepository.java:
```diff
@@ -2,6 +2,11 @@
     Iterable<Question> findByDeleted(boolean deleted);
+    Optional<Question> findFirstByTitleAndContents(String title, String contents);
+    Optional<Question> findByIdAndWriter(Long id, User writer);
+    void deleteByIdAndWriter(Long id, User writer);
```
이 메소드 없이 구현 진행해 본다.

javajigi commented on src/main/java/nextstep/domain/QuestionRepository.java:
```diff
@@ -2,6 +2,11 @@
+    Optional<Question> findFirstByTitleAndContents(String title, String contents);
+    Optional<Question> findByIdAndWriter(Long id, User writer);
+    void deleteByIdAndWriter(Long id, User writer);
 }
```
db 쿼리를 통해 로직을 구현하기 보다 애플리케이션 로직을 통해 같은 기능을 구현해 본다.

javajigi commented on src/main/java/nextstep/service/QnaService.java:
```diff
@@ -35,15 +36,20 @@ public Question create(User loginUser, Question question) {
-        // TODO 수정 기능 구현
-        return null;
+        updatedQuestion.setId(id);
+        updatedQuestion.writeBy(loginUser);
```
사용자가 전달한 updatedQuestion을 기반으로 기능을 구현하는 것은 문제가 발생할 가능성이 있다. 이 문제점이 무엇인지 찾아본다.
updatedQuestion을 이용하기 보다 db에서 Question을 조회한 후 이 객체를 업데이트하는 방식으로 구현한다.

javajigi commented on src/test/java/nextstep/web/UserAcceptanceTest.java:
```diff
@@ -73,12 +76,13 @@ public void update_no_login() throws Exception {
-                                                                            .build();
+        HttpEntity<MultiValueMap<String, Object>> request
+                = HtmlFormDataBuilder.urlEncodedForm()
+                                    .addParameter("_method", "put")
```
HtmlFormDataBuilder에 put() 메소드를 추가해 다음과 같이 구현하도록 하면 어떨까?

```
HtmlFormDataBuilder.urlEncodedForm().put()...
```

delete 요청도 같다.

javajigi reviewed (changes requested):
Acceptance Test 추가하는 것이 쉽지 않을텐데 구현 잘 했네요. 👍 
단, 로직 구현을 db 쿼리를 통해 구현하려고 하는데요. 그 보다 애플리케이션 로직을 통해 구현해볼 것을 추천해요.

elie-jehmi committed:
1. Question
	> 실제 db에서의 삭제가 아닌 deleted 상태를 이용하도록 수정.
2. HomeController)
	> delete false상태의 list 를 가져오도록 수정.
3. QuestionRepository, QuestionService
	> query base의 업무 로직을 제거, application 단에서 처리하도록 수정.
4. QnaController
	> exception 시 redirection home 으로 해주도록 수정.
5. UpdateForm.html
	> updateQuestion Post > Put으로 변경하여 Create/Update api 별도록 사용하도록 수정.
6. QnaAcceptanceTest, UserAcceptanceTst
	>변경사항에 따른 AcceptanceTest 수정.
7. HtmlFormDataBuilder
	> httpMethod 공통처리. (put/delete)

elie-jehmi committed:
Merge branch 'step2' of github.com:yesjehmi/java-qna-atdd into step2

javajigi commented on src/main/java/nextstep/domain/Question.java:
```diff
@@ -64,6 +64,10 @@ public void writeBy(User loginUser) {
         this.writer = loginUser;
     }
 
+    public boolean isWriter(User user) {
```
isOwner 메소드를 제공하고 있어 추가로 구현하지 않아도 됨

javajigi commented on src/main/java/nextstep/domain/Question.java:
```diff
@@ -77,6 +81,10 @@ public boolean isDeleted() {
         return deleted;
     }
 
+    public void delete() {
```
자신이 쓴 글인지에 대한 확인을 이 메소드에서 하는 것은 어떨까?

javajigi commented on src/main/java/nextstep/service/QnaService.java:
```diff
@@ -36,20 +37,28 @@ public Question create(User loginUser, Question question) {
+        question.setTitle(updatedQuestion.getTitle());
+        question.setContents(updatedQuestion.getContents());
+        question.writeBy(loginUser);
+        return questionRepository.save(question);
```
set 메소드 사용을 지양한다.
다음과 같이 구현하는 것은 어떨까?

```
question.update(loginUser, updatedQuestion);
```

questionRepository.save(..)와 같이 명시적으로 구현하지 않고 단순히 question.update(loginUser, updatedQuestion)만으로도 수정이 가능한지 테스트해 본다.

javajigi commented on src/main/java/nextstep/service/QnaService.java:
```diff
@@ -36,20 +37,28 @@ public Question create(User loginUser, Question question) {
+        }
+
+        question.delete();
+        questionRepository.save(question);
```
questionRepository.save(..)와 같이 명시적으로 구현하지 않고 단순히 question.delete()만으로도 수정이 가능한지 테스트해 본다.

yesjehmi replied:
음 이부분은 제가 잘 이해가 안가는데 다시한번 설명 부탁드려두될까요..
삭제를 위해서는 repository에서의 변경사항 저장 방식이 필요한데 
그걸 도메인 메소드 내부에서 처리하는게 어떤식으로 가능한건지 이해가 잘 가지 않습니다..

kang-hyungu replied:
jpa에서는 엔티티 변경사항을 데이터베이스에 자동으로 반영하는 변경 감지(dirty checking)이라는 기능이 있습니다.
트랜잭션 내에서 엔티티를 조회(꼭 조회해야합니다)해서 데이터를 변경하면 자동으로 update 쿼리를 수행합니다.
좀더 상세한 내용은 자바 ORM 표준 JPA 프로그래밍(저자 김영한) 102쪽 엔티티 수정을 참고하시기 바랍니다.

kang-hyungu replied:
update 쿼리가 동작하는지 확인하려면 questionRepository.save(question)를 주석처리하여 앱을 띄우고 콘솔창에서 어떤 쿼리가 실제로 실행되는지 확인해보시기 바랍니다.

kang-hyungu replied:
update 쿼리가 동작하는지 확인하려면 questionRepository.save(question)를 주석처리하여 앱을 띄우고 콘솔창에서 어떤 쿼리가 실제로 실행되는지 확인해보시기 바랍니다.

elie-jehmi replied:
와우 전에 들어본개념이었는데 jpa 많이 사용을 안해봐서 잊고있었네요.....ㅋㅋ
자세한 설명 감사합니다!!^__^

javajigi commented on src/main/java/nextstep/web/QnaController.java:
```diff
@@ -40,12 +39,25 @@ public String create(@LoginUser User user, Question question) {
+    public String update(@LoginUser User user, @PathVariable Long id,  Question question) {
+        try {
+            qnaService.update(user, id, question);
+            return "redirect:/"+id;
```
space convention 위반

javajigi reviewed (changes requested):
피드백 반영 잘 했네요. 👍 
추가로 개선했으면 하는 내용이 있어 피드백 좀 남겼어요.

elie-jehmi committed:
피드백 반영
1. Question 내부 중첩되는 메소드 제거 (isWriter)
2. QuestionService 에서 writer, loginUser check하는 로직 Question 내부로 보내기
3. space convention 통일

elie-jehmi committed:
step2 피드백 반영
1. QnaService > update, delete 이후 repository.save 제거.
2. 수정 완료 후 redirect url 변경
3. 수정 form 에서 put 방식 api 호출하도록 hidden method 추가

kang-hyungu reviewed (approved):
코드 수정하신 내용 확인했습니다.
구현 하느라 수고하셨습니다!
다음 단계 진행하시죠
